backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt

```diff
@@ -75,55 +74,113 @@ internal val IrClass.isArrayWithFixedSizeItems: Boolean
+        returnType.isVoidAsReturnType() -> TypeWithKind(null, TypeKind.VOID)
+        else -> TypeWithKind.fromType(returnType)
     }
+    ParameterIndex.DISPATCH_RECEIVER_INDEX -> TypeWithKind.fromType(dispatchReceiverParameter?.type)
+    ParameterIndex.EXTENSION_RECEIVER_INDEX -> TypeWithKind.fromType(extensionReceiverParameter?.type)
+    else -> TypeWithKind.fromType(this.valueParameters[index.unmap()].type)
 }
 
-private fun IrFunction.needBridgeToAt(target: IrFunction, index: Int)
-        = hasValueTypeAt(index) xor target.hasValueTypeAt(index)
+private fun IrFunction.needBridgeToAt(target: IrFunction, index: ParameterIndex)
+        = bridgeDirectionToAt(target, index).kind != BridgeDirectionKind.NONE
 
-internal fun IrFunction.needBridgeTo(target: IrFunction)
-        = (0..this.valueParameters.size + 2).any { needBridgeToAt(target, it) }
+@Suppress("EXPERIMENTAL_FEATURE_WARNING")
+private inline class ParameterIndex(val index: Int) {
+    companion object {
+        val RETURN_INDEX = ParameterIndex(0)
+        val DISPATCH_RECEIVER_INDEX = ParameterIndex(1)
+        val EXTENSION_RECEIVER_INDEX = ParameterIndex(2)
 
-internal enum class BridgeDirection {
-    NOT_NEEDED,
-    FROM_VALUE_TYPE,
-    TO_VALUE_TYPE
+        fun map(index: Int) = ParameterIndex(index + 3)
+
+        fun allParametersCount(irFunction: IrFunction) = irFunction.valueParameters.size + 3
+
+        inline fun forEachIndex(irFunction: IrFunction, block: (ParameterIndex) -> Unit) =
+                (0 until allParametersCount(irFunction)).forEach { block(ParameterIndex(it)) }
+    }
+
+    fun unmap() = index - 3
 }
 
-private fun IrFunction.bridgeDirectionToAt(target: IrFunction, index: Int)
-       = when {
-            hasValueTypeAt(index) && target.hasReferenceAt(index) -> BridgeDirection.FROM_VALUE_TYPE
-            hasReferenceAt(index) && target.hasValueTypeAt(index) -> BridgeDirection.TO_VALUE_TYPE
-            else -> BridgeDirection.NOT_NEEDED
+internal fun IrFunction.needBridgeTo(target: IrFunction): Boolean {
+    ParameterIndex.forEachIndex(this) {
+        if (needBridgeToAt(target, it)) return true
+    }
+    return false
+}
+
+internal enum class BridgeDirectionKind {
+    NONE,
+    BOX,
+    UNBOX
+}
+
+internal data class BridgeDirection(val irClass: IrClass?, val kind: BridgeDirectionKind) {
+    companion object {
+        val NONE = BridgeDirection(null, BridgeDirectionKind.NONE)
+    }
+}
+
+private fun IrFunction.bridgeDirectionToAt(overriddenFunction: IrFunction, index: ParameterIndex): BridgeDirection {
+    val kind = typeWithKindAt(index).kind
+    val (irClass, otherKind) = overriddenFunction.typeWithKindAt(index)
+    if ((kind == TypeKind.VOID || kind == TypeKind.REFERENCE) && otherKind != kind)
+        return BridgeDirection(irClass?.erasure(), BridgeDirectionKind.UNBOX)
+    if (kind == TypeKind.VALUE_TYPE && otherKind == TypeKind.REFERENCE)
+        return BridgeDirection(null /* Erase to [Any?] */, BridgeDirectionKind.BOX)
```

Comment:
I'm not sure that `kind == TypeKind.VALUE_TYPE && otherKind == TypeKind.VOID` case is impossible.

Consider the following example:
```kotlin
open class Base1<T : Int> {
    fun foo(): T { 
        TODO()
    }
}

interface Base2 {
    fun foo(): Nothing
}

class Derived : Base1<Nothing>(), Base2 {
}
```

It requires a bridge and probably meets the condition above.

Reply:
Indeed.